File: apps/api/src/properties/property-nesting.ts
import { HttpStatus } from '@nestjs/common';
import { PropertyType } from '@prisma/client';
import { NodeScopeException } from '../common/filters/global-exception.filter';

// Parent type → the child types it may contain (spec §5).
export const ALLOWED_CHILDREN: Record<PropertyType, PropertyType[]> = {
  SITE: ['SITE', 'BUILDING', 'AREA'],
  BUILDING: ['FLOOR', 'AREA'],
  FLOOR: ['AREA'],
  AREA: ['AREA'],
};

/** Throws PROP_002 if `childType` may not sit under `parentType` (null parent = root, must be SITE). */
export function assertValidNesting(parentType: PropertyType | null, childType: PropertyType): void {
  if (parentType === null) {
    if (childType !== 'SITE') {
      throw new NodeScopeException('PROP_002', 'INVALID_PARENT_TYPE', HttpStatus.UNPROCESSABLE_ENTITY);
    }
    return;
  }
  if (!ALLOWED_CHILDREN[parentType].includes(childType)) {
    throw new NodeScopeException('PROP_002', 'INVALID_PARENT_TYPE', HttpStatus.UNPROCESSABLE_ENTITY);
  }
}

File: src/Modules/Inventory/NodeScope.Modules.Inventory.Domain/PropertyNesting.cs
using NodeScope.Platform.Abstractions;

namespace NodeScope.Modules.Inventory.Domain;

/// <summary>The hierarchy rules from Node's <c>property-nesting.ts</c> (spec §5).</summary>
public static class PropertyNesting
{
    private static readonly Dictionary<PropertyType, PropertyType[]> AllowedChildren = new()
    {
        [PropertyType.Site] = [PropertyType.Site, PropertyType.Building, PropertyType.Area],
        [PropertyType.Building] = [PropertyType.Floor, PropertyType.Area],
        [PropertyType.Floor] = [PropertyType.Area],
        [PropertyType.Area] = [PropertyType.Area],
    };

    /// <summary>
    /// Throws <c>PROP_002</c> when <paramref name="childType"/> may not sit under
    /// <paramref name="parentType"/> (null parent = root, which must be a SITE).
    /// </summary>
    public static void AssertValidNesting(PropertyType? parentType, PropertyType childType)
    {
        if (parentType is null)
        {
            if (childType != PropertyType.Site)
            {
                throw InvalidParentType();
            }

            return;
        }

        if (!AllowedChildren[parentType.Value].Contains(childType))
        {
            throw InvalidParentType();
        }
    }

    private static ApiException InvalidParentType() => new("PROP_002", "INVALID_PARENT_TYPE", 422);
}
